File: tests/refs/kruskal_mst_weight.cpp
// Reference C++ driver for tests/cases/kruskal_mst_weight.json. Union-find.
#include <algorithm>
#include <iostream>
#include <numeric>
#include <sstream>
#include <string>
#include <vector>

struct DSU {
    std::vector<int> p, r;
    DSU(int n) : p(n), r(n, 0) { std::iota(p.begin(), p.end(), 0); }
    int find(int x) { while (p[x] != x) { p[x] = p[p[x]]; x = p[x]; } return x; }
    bool unite(int a, int b) {
        a = find(a); b = find(b); if (a == b) return false;
        if (r[a] < r[b]) std::swap(a, b);
        p[b] = a; if (r[a] == r[b]) ++r[a]; return true;
    }
};

long long kruskal(int V, std::vector<std::tuple<int,int,long long>> edges) {
    std::sort(edges.begin(), edges.end(), [](auto& a, auto& b) { return std::get<2>(a) < std::get<2>(b); });
    DSU d(V); long long total = 0; int used = 0;
    for (auto& [u, v, w] : edges) if (d.unite(u, v)) { total += w; if (++used == V - 1) break; }
    return total;
}

int main() {
    std::string line, name;
    int V = 0, edgesRemain = 0; bool inEdges = false;
    std::vector<std::tuple<int,int,long long>> edges;
    long long expected = 0;
    bool hi = false, he = false; int p = 0, f = 0;
    auto run = [&]() {
        if (name.empty() || !hi || !he) return;
        long long got = kruskal(V, edges);
        if (got == expected) { std::cout << "PASS kruskal_mst_weight :: " << name << "\n"; ++p; }
        else { std::cout << "FAIL kruskal_mst_weight :: " << name << "\n  expected: " << expected << "\n  got: " << got << "\n"; ++f; }
        name.clear(); edges.clear(); hi = he = false; inEdges = false;
    };
    while (std::getline(std::cin, line)) {
        if (line.empty()) { run(); continue; }
        std::istringstream iss(line); std::string tag; iss >> tag;
        if (tag == "CASE") { iss >> name; }
        else if (tag == "V") { iss >> V; }
        else if (tag == "EDGES") { iss >> edgesRemain; inEdges = true; edges.clear(); if (edgesRemain == 0) { hi = true; inEdges = false; } }
        else if (tag == "INT") { iss >> expected; he = true; }
        else if (inEdges) {
            std::istringstream is2(line); int u, v; long long w; is2 >> u >> v >> w;
            edges.push_back({u, v, w}); --edgesRemain;
            if (edgesRemain == 0) { hi = true; inEdges = false; }
        }
    }
    run();
    std::cout << "TOTAL: " << p << " passed, " << f << " failed\n";
    return f == 0 ? 0 : 1;
}

File: tests/refs/kruskal_mst_weight.rs
// Reference Rust driver for tests/cases/kruskal_mst_weight.json.
use std::io::{self, BufRead};

struct DSU { p: Vec<usize>, r: Vec<i32> }
impl DSU {
    fn new(n: usize) -> Self { DSU { p: (0..n).collect(), r: vec![0; n] } }
    fn find(&mut self, mut x: usize) -> usize {
        while self.p[x] != x { self.p[x] = self.p[self.p[x]]; x = self.p[x]; }
        x
    }
    fn unite(&mut self, a: usize, b: usize) -> bool {
        let (mut a, mut b) = (self.find(a), self.find(b));
        if a == b { return false; }
        if self.r[a] < self.r[b] { std::mem::swap(&mut a, &mut b); }
        self.p[b] = a;
        if self.r[a] == self.r[b] { self.r[a] += 1; }
        true
    }
}

fn kruskal(v: usize, mut edges: Vec<(i32, i32, i64)>) -> i64 {
    edges.sort_by_key(|e| e.2);
    let mut d = DSU::new(v);
    let mut total = 0i64; let mut used = 0usize;
    for (u, w, ww) in edges {
        if d.unite(u as usize, w as usize) {
            total += ww;
            used += 1;
            if used == v.saturating_sub(1) { break; }
        }
    }
    total
}

fn main() {
    let stdin = io::stdin();
    let mut name = String::new();
    let mut v_in: usize = 0;
    let mut edges: Vec<(i32, i32, i64)> = Vec::new();
    let mut edges_remain: usize = 0; let mut in_edges = false;
    let mut expected: i64 = 0;
    let mut hi = false; let mut he = false;
    let mut p = 0u32; let mut f = 0u32;
    let flush = |name: &mut String, v_in: usize, edges: &mut Vec<(i32, i32, i64)>, expected: i64,
                 hi: &mut bool, he: &mut bool, in_edges: &mut bool, p: &mut u32, f: &mut u32| {
        if !name.is_empty() && *hi && *he {
            let got = kruskal(v_in, edges.clone());
            if got == expected { println!("PASS kruskal_mst_weight :: {}", name); *p += 1; }
            else { println!("FAIL kruskal_mst_weight :: {}\n  expected: {}\n  got: {}", name, expected, got); *f += 1; }
        }
        name.clear(); edges.clear(); *hi = false; *he = false; *in_edges = false;
    };
    for line in stdin.lock().lines() {
        let line = line.unwrap_or_default();
        if line.is_empty() { flush(&mut name, v_in, &mut edges, expected, &mut hi, &mut he, &mut in_edges, &mut p, &mut f); continue; }
        let mut parts = line.split_whitespace();
        let tag = parts.next().unwrap_or("");
        match tag {
            "CASE" => { name = parts.next().unwrap_or("").to_string(); }
            "V" => { v_in = parts.next().and_then(|s| s.parse().ok()).unwrap_or(0); }
            "EDGES" => {
                edges_remain = parts.next().and_then(|s| s.parse().ok()).unwrap_or(0);
                in_edges = true; edges.clear();
                if edges_remain == 0 { hi = true; in_edges = false; }
            }
            "INT" => { expected = parts.next().and_then(|s| s.parse().ok()).unwrap_or(0); he = true; }
            _ => {
                if in_edges {
                    let u: i32 = tag.parse().unwrap_or(0);
                    let w: i32 = parts.next().and_then(|s| s.parse().ok()).unwrap_or(0);
                    let ww: i64 = parts.next().and_then(|s| s.parse().ok()).unwrap_or(0);
                    edges.push((u, w, ww));
                    edges_remain -= 1;
                    if edges_remain == 0 { hi = true; in_edges = false; }
                }
            }
        }
    }
    flush(&mut name, v_in, &mut edges, expected, &mut hi, &mut he, &mut in_edges, &mut p, &mut f);
    println!("TOTAL: {} passed, {} failed", p, f);
    std::process::exit(if f == 0 { 0 } else { 1 });
}
